### PR TITLE
[c10d] add fake pg necessary collectives

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102238
* #102237

This PR adds fake pg necessary collectives to enable e2e FSDP run
with out multiprocess or multithreading

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1125,10 +1125,7 @@ def init_process_group(
         # these barriers may be unnecessary, as proved by a green CI after
         # removal. An environment variable `TORCH_DIST_INIT_BARRIER` has been
         # added which, when set to 0, will disable these barriers.
-        if backend == "fake":
-            # Fake process group doesn't need barrier
-            pass
-        elif backend == Backend.MPI:
+        if backend == Backend.MPI:
```
Comment (ezyang) [inline torch/distributed/distributed_c10d.py:1128]: Nyehh, I think I preferred not mucking about with the environment variable instead haha.

**Review by ezyang — APPROVED** [review]

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/distributed_c10d.py:1128]: Yeah I kinda feel bad for this too, it was easier for me to leverage the previous work that uses a environ knob to control barrier init, let me think about it, there should be some better way for custom backend.

Comment (wanchaol) [reply torch/distributed/distributed_c10d.py:1128]: landing this first to unblock dev efficiency and will address this in a follow up PR

Comment (wanchaol) [reply torch/distributed/distributed_c10d.py:1128]: follow up PR here: https://github.com/pytorch/pytorch/pull/102343

**Pull request closed without merging**